### PR TITLE
Ports MC deletion recovery

### PR DESCRIPTION
**Pull request by ike709**
## About The Pull Request

https://github.com/tgstation/tgstation/pull/60846

It means we can delete the MC with 100% impunity whenever we desire and everything will be 100% fine. What could possibly go wrong?

On a more serious note, we need to do a fair bit of testing to see how this interacts with auxmos.

## Changelog
:cl: Gamer025 and ike709
code: Failsafe is more robust now and the game can actually recover from a deleted MC
code: Albeit far from perfect, the game will be fairly playable if the Master Controller dies. It beats the alternative by a lot if there's no admin online to reboot the server.
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/economy.dm [hunk 1]
```diff
@@ -27,6 +27,11 @@ SUBSYSTEM_DEF(economy)
 		new /datum/bank_account/department(A, budget_to_hand_out)
 	return ..()
 
+/datum/controller/subsystem/economy/Recover()
+	generated_accounts = SSeconomy.generated_accounts
+	bank_accounts_by_id = SSeconomy.bank_accounts_by_id
```
Comment (ike709) [inline code/controllers/subsystem/economy.dm]:
```suggestion
```

File: code/controllers/failsafe.dm [hunk 2]
```diff
@@ -33,6 +33,22 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+		while (defcon > 0)
+			var/recovery_result = emergency_loop()
+			if (recovery_result == 1) //Exit emergency loop and delete self if it was able to recover MC
+				break
+			else if (defcon == 1) //Exit Failsafe if we weren't able to recover the MC in the last stage
+				log_game("FailSafe: Failed to recover MC while in emergency state. Failsafe exiting.")
+				message_admins(span_boldannounce("Failsafe failed criticaly while trying to recreate broken MC. Please manually fix the MC or reboot the server. Failsafe exiting now."))
+				message_admins(span_boldannounce("You can try manually calling these two procs:."))
+				message_admins(span_boldannounce("/proc/recover_all_SS_and_recreate_master: Most stuff should still function but expect instability/runtimes/broken stuff."))
+				message_admins(span_boldannounce("/proc/delete_all_SS_and_recreate_master: Most stuff will be broken but basic stuff like movement and chat should still work."))
```
Comment (ike709) [inline code/controllers/failsafe.dm]:
```suggestion
				message_admins(<span class='boldannounce'>"Failsafe failed criticaly while trying to recreate broken MC. Please manually fix the MC or reboot the server. Failsafe exiting now.</span>")
				message_admins(<span class='boldannounce'>"You can try manually calling these two procs:.</span>")
				message_admins(<span class='boldannounce'>"/proc/recover_all_SS_and_recreate_master: Most stuff should still function but expect instability/runtimes/broken stuff.</span>")
				message_admins(<span class='boldannounce'>"/proc/delete_all_SS_and_recreate_master: Most stuff will be broken but basic stuff like movement and chat should still work.</span>")
```

File: code/controllers/failsafe.dm [hunk 3]
```diff
@@ -92,6 +108,57 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+			. = Recreate_MC()
+
+	if (. == 1) //We were able to create a new master
+		master_iteration = 0
+		SSticker.Recover(); //Recover the ticket system so the Masters runlevel gets set
+		Master.Initialize(10, FALSE, TRUE) //Need to manually start the MC, normally world.new would do this
+		to_chat(GLOB.admins, span_adminnotice("Failsafe recovered MC while in emergency state [defcon_pretty()]"))
+	else
+		log_game("FailSafe: Failsafe in emergency state and was unable to recreate MC while in defcon state [defcon_pretty()].")
+		message_admins(span_boldannounce("Failsafe in emergency state and master down, trying to recreate MC while in defcon level [defcon_pretty()] failed."))
```
Comment (ike709) [inline code/controllers/failsafe.dm]:
```suggestion
		to_chat(GLOB.admins, "<span class='adminnotice'>Failsafe recovered MC while in emergency state [defcon_pretty()]</span>")
	else
		log_game("FailSafe: Failsafe in emergency state and was unable to recreate MC while in defcon state [defcon_pretty()].")
		message_admins("<span class='boldannounce'>Failsafe in emergency state and master down, trying to recreate MC while in defcon level [defcon_pretty()] failed.</span>")
```

File: code/controllers/failsafe.dm [hunk 4]
```diff
@@ -92,6 +108,57 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+	for(var/global_var in global.vars)
+		if (istype(global.vars[global_var], /datum/controller/subsystem))
+			del(global.vars[global_var])
+	. = Recreate_MC()
+	if (. == 1) //We were able to create a new master
+		SSticker.Recover(); //Recover the ticket system so the Masters runlevel gets set
+		Master.Initialize(10, FALSE, TRUE) //Need to manually start the MC, normally world.new would do this
+		to_chat(GLOB.admins, span_adminnotice("MC successfully recreated after deleting and recreating all subsystems!"))
+	else
+		message_admins(span_boldannounce("Failed to create new MC!"))
```
Comment (ike709) [inline code/controllers/failsafe.dm]:
```suggestion
		to_chat(GLOB.admins, <span class='adminnotice'>"MC successfully recreated after deleting and recreating all subsystems!</span>")
	else
		message_admins("<span class='boldannounce'>Failed to create new MC!</span>")
```

File: code/controllers/failsafe.dm [hunk 5]
```diff
@@ -92,6 +108,57 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+/proc/delete_all_SS_and_recreate_master()
+	del(Master)
+	for(var/global_var in global.vars)
+		if (istype(global.vars[global_var], /datum/controller/subsystem))
+			del(global.vars[global_var])
+	. = Recreate_MC()
+	if (. == 1) //We were able to create a new master
+		SSticker.Recover(); //Recover the ticket system so the Masters runlevel gets set
+		Master.Initialize(10, FALSE, TRUE) //Need to manually start the MC, normally world.new would do this
+		to_chat(GLOB.admins, <span class='adminnotice'>"MC successfully recreated after deleting and recreating all subsystems!</span>")
```
Comment (ike709) [inline code/controllers/failsafe.dm]:
```suggestion
		to_chat(GLOB.admins, "<span class='adminnotice'>MC successfully recreated after deleting and recreating all subsystems!</span>")
```

File: code/controllers/failsafe.dm [hunk 6]
```diff
@@ -92,6 +108,57 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+	sortTim(subsytem_types, /proc/cmp_subsystem_init)
+	for(var/I in subsytem_types)
+		new I
+	. = Recreate_MC()
+	if (. == 1) //We were able to create a new master
+		SSticker.Recover(); //Recover the ticket system so the Masters runlevel gets set
+		Master.Initialize(10, FALSE, TRUE) //Need to manually start the MC, normally world.new would do this
+		to_chat(GLOB.admins, span_adminnotice("MC successfully recreated after recovering all subsystems!"))
+	else
+		message_admins(span_boldannounce("Failed to create new MC!"))
```
Comment (ike709) [inline code/controllers/failsafe.dm]:
```suggestion
		to_chat(GLOB.admins, "<span class='adminnotice'>MC successfully recreated after recovering all subsystems!</span>")
	else
		message_admins("<span class='boldannounce'>Failed to create new MC!</span>")
```

File: code/controllers/failsafe.dm [hunk 7]
```diff
@@ -33,6 +33,22 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+		while (defcon > 0)
+			var/recovery_result = emergency_loop()
+			if (recovery_result == 1) //Exit emergency loop and delete self if it was able to recover MC
+				break
+			else if (defcon == 1) //Exit Failsafe if we weren't able to recover the MC in the last stage
+				log_game("FailSafe: Failed to recover MC while in emergency state. Failsafe exiting.")
+				message_admins(<span class='boldannounce'>"Failsafe failed criticaly while trying to recreate broken MC. Please manually fix the MC or reboot the server. Failsafe exiting now.</span>")
+				message_admins(<span class='boldannounce'>"You can try manually calling these two procs:.</span>")
+				message_admins(<span class='boldannounce'>"/proc/recover_all_SS_and_recreate_master: Most stuff should still function but expect instability/runtimes/broken stuff.</span>")
+				message_admins(<span class='boldannounce'>"/proc/delete_all_SS_and_recreate_master: Most stuff will be broken but basic stuff like movement and chat should still work.</span>")
```
Comment (ike709) [inline code/controllers/failsafe.dm]:
```suggestion
				message_admins("<span class='boldannounce'>Failsafe failed criticaly while trying to recreate broken MC. Please manually fix the MC or reboot the server. Failsafe exiting now.</span>")
				message_admins("<span class='boldannounce'>You can try manually calling these two procs:.</span>")
				message_admins("<span class='boldannounce'>/proc/recover_all_SS_and_recreate_master: Most stuff should still function but expect instability/runtimes/broken stuff.</span>")
				message_admins("<span class='boldannounce'>/proc/delete_all_SS_and_recreate_master: Most stuff will be broken but basic stuff like movement and chat should still work.</span>")
```

**Comment by ike709** [comment]
Oh my god it works.

**Comment by utilbird** [comment]
Thanks to the excellent work of our best and brightest, I can now delete the master controller with a clear conscience

**Comment by Archanial** [comment]
doesn't work lol

**Comment by ike709** [comment]
@Archanial TG's SStimer can't recover, but I got it working.

**Comment by ike709** [comment]
Testmerged for another round. The only real issue noted was the mining shuttle console being broken by MC deletion & recovery.

Also got a report of exploration consoles generally breaking.

But frankly that's such a nonissue in the scope of the entire MC dying that it definitely shouldn't block merge. Would be nice if @PowerfulBacon could look into it though.

File: code/controllers/subsystem/zclear.dm [hunk 8]
```diff
@@ -37,6 +37,15 @@ SUBSYSTEM_DEF(zclear)
...
 	ignored_atoms = typecacheof(list(/mob/dead, /mob/camera, /mob/dview, /atom/movable/lighting_object, /obj/effect/abstract/mirage_holder))
 
+/datum/controller/subsystem/zclear/Recover()
+	autowipe = SSzclear.autowipe
+	free_levels = SSzclear.free_levels
+	processing_levels = SSzclear.processing_levels
+	ignored_atoms = SSzclear.ignored_atoms
+	nullspaced_mobs = SSzclear.nullspaced_mobs
+	docking_levels = SSzclear.docking_levels
+	announced_zombie_levels = SSzclear.announced_zombie_levels
```
Comment (ike709) [inline code/controllers/subsystem/zclear.dm]:
```suggestion
	autowipe |= SSzclear.autowipe
	free_levels |= SSzclear.free_levels
	processing_levels |= SSzclear.processing_levels
	ignored_atoms |= SSzclear.ignored_atoms
	nullspaced_mobs |= SSzclear.nullspaced_mobs
	docking_levels |= SSzclear.docking_levels
	announced_zombie_levels |= SSzclear.announced_zombie_levels
```

**Comment by ike709** [comment]
Exploration maybe fixed.

**Comment by ike709** [comment]
This PR isn't perfect, and tbh I'll probably never invest the effort to make it perfect, but it will keep the round Mostly Playable(TM) if the MC dies when an admin isn't around to fix/reboot things.

And frankly that's good enough for me.

**Review by Archanial — APPROVED** [review]

**Pull request merged**